source/fe/mapping_fe.cc

```diff
@@ -0,0 +1,1385 @@
+  const std::vector<Point<dim> > &unit_points,
+  typename MappingFE<dim, spacedim>::InternalData &data) const
+{
+  const unsigned int n_points=unit_points.size();
+
+  if (data.shape_values.size()!=0 || data.shape_derivatives.size()!=0)
+    for (unsigned int point=0; point<n_points; ++point)
+      {
+        if (data.shape_values.size()!=0)
+          for (unsigned int i=0; i<data.n_shape_functions; ++i)
+            data.shape(point, i) = fe->shape_value(i, unit_points[point]);
+
+        if (data.shape_derivatives.size()!=0)
+          for (unsigned int i=0; i<data.n_shape_functions; ++i)
+            data.derivative(point, i) = fe->shape_grad(i, unit_points[point]);
+
+        if (data.shape_second_derivatives.size()!=0)
+          for (unsigned int i=0; i<data.n_shape_functions; ++i)
+            data.second_derivative(point, i) = fe->shape_grad_grad(i, unit_points[point]);
+      }
+}
+
+
+template<int dim, int spacedim, class DH, class VECTOR>
+UpdateFlags
+MappingFE<dim,spacedim,DH,VECTOR>::update_once (const UpdateFlags in) const
+{
+  UpdateFlags out = UpdateFlags(in & (update_transformation_values
+                                      | update_transformation_gradients));
+
+  // Shape function values
+  if (in & update_quadrature_points)
+    out |= update_transformation_values;
+
+  // Shape function gradients
+  if (in & (update_covariant_transformation
+            | update_contravariant_transformation
+            | update_JxW_values
+            | update_boundary_forms
+            | update_normal_vectors
+            | update_jacobians
+            | update_jacobian_grads
+            | update_inverse_jacobians))
+    out |= update_transformation_gradients;
+
+  return out;
+}
+
+
+
+template<int dim, int spacedim, class DH, class VECTOR>
+UpdateFlags
+MappingFE<dim,spacedim,DH,VECTOR>::update_each (const UpdateFlags in) const
+{
+  // Select flags of concern for the
+  // transformation.
+  UpdateFlags out = UpdateFlags(in & (update_quadrature_points
+                                      | update_covariant_transformation
+                                      | update_contravariant_transformation
+                                      | update_JxW_values
+                                      | update_boundary_forms
+                                      | update_normal_vectors
+                                      | update_volume_elements
+                                      | update_jacobians
+                                      | update_jacobian_grads
+                                      | update_inverse_jacobians));
+
+  // add flags if the respective
+  // quantities are necessary to
+  // compute what we need. note that
+  // some flags appear in both
+  // conditions and in subsequents
+  // set operations. this leads to
+  // some circular logic. the only
+  // way to treat this is to
+  // iterate. since there are 4
+  // if-clauses in the loop, it will
```

Comment:
There are actually five `if` clauses. Does this affect the need to run 4 iterations?